samples/wasi-threads/wasm-apps/CMakeLists.txt

```diff
@@ -29,6 +29,8 @@ function (compile_sample SOURCE_FILE)
     LINKER:--export=__data_end
     LINKER:--shared-memory,--max-memory=1966080
     LINKER:--export=wasi_thread_start
+    LINKER:--export=malloc
+    LINKER:--export=free
```

Comment:
why?

Reply:
Because I was getting this error without the additional exports:
```
[09:52:39:510 - 101150580]: Memory instantiate success.
[09:52:39:614 - 101150580]: Error: app heap is corrupted, if the wasm file is compiled by wasi-sdk-12.0 or higher version, please add -Wl,--export=malloc -Wl,--export=free to export malloc and free functions. If it is compiled by asc, please add --exportRuntime to export the runtime helpers.
[09:52:39:618 - 101150580]: thread manager error: failed to allocate aux stack space for new thread
[09:52:39:620 - 101150580]: Failed to spawn a new thread
[09:52:39:623 - 101150580]: /Users/eloparco/dev/forks/wasm-micro-runtime/core/iwasm/interpreter/wasm_interp_fast.c, line 3959, meet an exception Exception: app heap corrupted
Exception: app heap corrupted
```

Reply:
the aux stack allocation should be disabled for wasi-threads because it's libc's responsibility to allocate stack for new threads.

no_pthread.c seems broken. its wasi_thread_start should either update the stack pointer global or avoid using C stack.

Reply:
> the aux stack allocation should be disabled for wasi-threads because it's libc's responsibility to allocate stack for new threads.

https://github.com/bytecodealliance/wasm-micro-runtime/blob/dev/wasi_threads/core/config.h#L172

> no_pthread.c seems broken. its wasi_thread_start should either update the stack pointer global or avoid using C stack.

can you elaborate on that and give me some hints on what would need to be changed? my knowledge on the topic is quite limited

Reply:
According to [this](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/doc/memory_tune.md),

> app heap and libc heap: the heap to allocate memory for wasm app, note that app heap is created only when the malloc/free functions (or __new/__release functions for AssemblyScript) are not exported and runtime can not detect the libc heap.

Reply:
I've read that one already, but if I don't export malloc and free I'm getting the corruption error

Reply:
> > the aux stack allocation should be disabled for wasi-threads because it's libc's responsibility to allocate stack for new threads.
> 
> https://github.com/bytecodealliance/wasm-micro-runtime/blob/dev/wasi_threads/core/config.h#L172

https://github.com/bytecodealliance/wasm-micro-runtime/pull/1799#discussion_r1045454952

> 
> > no_pthread.c seems broken. its wasi_thread_start should either update the stack pointer global or avoid using C stack.
> 
> How do I use the app heap? if I don't export the malloc and free functions I get `Error: app heap is corrupted`

wamr doesn't need to (and shouldn't) use the heap (app heap or libc heap) at least for this purpose (stack allocation for wasi-threads) because the stack is managed inside the module itself. (usually by libc)

Reply:
Threads implementation in WAMR is not fully completed yet. One of the things is to *always* (regardless of compilation flags) use on-heap aux stack allocation. I have change for that already but didn't test it yet; I'll make the change public beginning of next month when I'm back from holidays, until then I'd suggest waiting with merging this PR.